pkg/kubeapiserver/options/authentication.go

```diff
@@ -269,6 +269,25 @@ func (o *BuiltInAuthenticationOptions) Validate() []error {
 
 	if o.RequestHeader != nil {
 		allErrors = append(allErrors, o.RequestHeader.Validate()...)
+		if o.ClientCert != nil {
+			// Enforce 'requestheader-allowed-names' check when 'requestheader-client-ca-file'
+			// and 'client-ca-file' are the same and not empty, to prevent unauthorized privilege escalation.
+			clientCA := strings.TrimSpace(o.ClientCert.ClientCA)
+			requestHeaderClientCA := strings.TrimSpace(o.RequestHeader.ClientCAFile)
+			if len(clientCA) != 0 && clientCA == requestHeaderClientCA {
```

Comment:
comparing the filenames is not really what we want, this needs to load both CA pools and check if any CA is in both client CA and requestheader CA list:

```go
if len(o.RequestHeader.AllowedNames) == 0 && o.ClientCert != nil && len(o.ClientCert.ClientCA) > 0 && len(o.RequestHeader.ClientCAFile) > 0 {
  clientCACerts, _ := cert.CertsFromFile(o.ClientCert.ClientCA)
  requestHeaderCACerts, _ := cert.CertsFromFile(o.RequestHeader.ClientCAFile)
  ... check if any cert is in both clientCACerts and requestHeaderCACerts
}
```